lib/pure/os.nim

```diff
@@ -2088,6 +2095,221 @@ iterator walkDir*(dir: string; relative=false): tuple[kind: PathComponent, path:
+               else: dir / extractFilename(getFilename(fdat))
+      sNoErrors(k, xx)
+
+    var f: WIN32_FIND_DATA
+    var h: Handle = findFirstFile(dir / "*", f)
+    defer:
+      if h != -1:
+        findClose(h)
+    let status =
+      if h == -1:
+        case getLastError()
+        of ERROR_PATH_NOT_FOUND: odNotFound
+        of ERROR_DIRECTORY: odNotDir
+        of ERROR_ACCESS_DENIED: odAccessDenied
+        else: odUnknownError
+      else: odOpenOk
+    step = sOpenDir(status, outDir)
+
+    var firstStep = true
+    while true:
+      if not skip:
+        yield step
+      if h == -1 or step.kind == wsInterrupted:
+        break
+      if firstStep:  # use file obtained by findFirstFile
+        firstStep = false
+      else:  # get next file
+        if findNextFile(h, f) == 0'i32:
+          let errCode = getLastError()
+          if errCode == ERROR_NO_MORE_FILES:
+            break  # normal end, yielding nothing
+          else:
+            skip = false
+            step = sGetError(wsInterrupted, outDir)
+            continue
+      skip = skipFindData(f)
+      if not skip:
+        step = resolveFile(f, relative)
+
+  else:
+    proc resolveFile(x: ptr Dirent, path: string, y: string): WalkStep =
+      var k = pcFile
+      when defined(linux) or defined(macosx) or
+           defined(bsd) or defined(genode) or defined(nintendoswitch):
+        if x.d_type != DT_UNKNOWN:
+          if x.d_type == DT_DIR:
+            k = pcDir
+            return sNoErrors(k, y)
+          elif x.d_type == DT_LNK:
+            errno = 0
+            if dirExists(path): k = pcLinkToDir
+            else: k = pcLinkToFile
+            # check error in dirExists
+            if errno == 0:
+              return sNoErrors(k, y)
+            else:
+              return sGetError(wsEntryBad, y)
+          else:
+            return sNoErrors(k, y)
+      # case of d_type==DT_UNKNOWN: some filesystems don't report
+      # the entry type; one should fallback to lstat
+      var s: Stat
+      if lstat(path, s) < 0'i32:
+        return sGetError(wsEntryBad, y)
+      else:
+        errno = 0
+        if S_ISDIR(s.st_mode):
+          k = pcDir
+        elif S_ISLNK(s.st_mode):
+          k = getSymlinkFileKind(path)
+        # check error in getSymlinkFileKind
+        if errno == 0:
+          return sNoErrors(k, y)
+        else:
+          return sGetError(wsEntryBad, y)
+
+    var d: ptr DIR = opendir(dir)
+    defer:
+      if d != nil:
+        discard closedir(d)
+    let status =
+      if d == nil:
+          if errno == ENOENT: odNotFound
+          elif errno == ENOTDIR: odNotDir
+          elif errno == EACCES: odAccessDenied
+          else: odUnknownError
+      else: odOpenOk
+    step = sOpenDir(status, outDir)
+
+    while true:
+      if not skip:
+        yield step
+      if d == nil or step.kind == wsInterrupted:
+        break
+      let errnoSave = errno
+      defer: errno = errnoSave
+      errno = 0
```

Comment:
Suspicious code.

Reply:
fixed, preserving errno is done more correct way, without defer.